sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/WriterContainer.scala

```diff
@@ -324,6 +324,8 @@ private[sql] class DynamicPartitionWriterContainer(
 
   private def bucketIdExpression: Option[Expression] = for {
     BucketSpec(numBuckets, _, _) <- bucketSpec
+  // Use `Pmod` to match the behaviour of `Utils.nonNegativeMod`, which is the function used to
```

Comment:
nonNegativeMod seems to be used in core places so making it depend on a SQL class seems wrong. I was just suggesting that the int version of pmod just call utils.nonNegativeMod and you add the comment saying changing this is bad.